encryptor.py

```diff
@@ -0,0 +1,293 @@
+    return input_text
+
+
+def vernam_encode_decode(args, input_text):
+    if not args.key.isalpha():
+        exit("The key of vigenere's cipher is a word")
+    size_block = 32
+    for i in range(len(input_text)):
+        if 0 <= ord(input_text[i]) - ord('a') < size_block or 0 <= ord(input_text[i]) - ord('A') < size_block:
+            key_elem = args.key[i % len(args.key)].lower()
+            elem = chr(((ord(input_text[i]) - ord('a')) ^ (ord(key_elem) - ord('a'))) + ord('a'))
+            input_text = input_text[:i] + elem + input_text[i + 1:]
+    return input_text
+
+
+def caesar_decode(args, input_text):
+    if not args.key.isdigit():
+        sys.exit("The key of сaesar's cipher is a number")
+    args.key = str(SIZE_ALPHABET - int(args.key))
+    return caesar_encode(args, input_text)
+
+
+def vigenere_decode(args, input_text):
+    if not args.key.isalpha():
+        exit("The key of vigenere's cipher is a word")
+    args.key = args.key.lower()
+    for i in range(len(args.key)):
+        args.key = args.key[:i] + chr((ord('a') + SIZE_ALPHABET - ord(args.key[i])) % SIZE_ALPHABET + ord('a'))\
+                   + args.key[i + 1:]
+    return vigenere_encode(args, input_text)
+
+
+def train(args):
+    input_text = read_from_file(args.text_file)
+    input_text = parc(input_text)
+    count_symb = len(input_text)
+    dict_symb = {chr(ord('a') + i): 0 for i in range(SIZE_ALPHABET)}
+    for elem in input_text:
+        dict_symb[elem] += 1
+    if count_symb == 0:
+        exit("The input file does not contain any characters")
+    else:
+        for key in dict_symb.keys():
+            dict_symb[key] = dict_symb[key] / count_symb
+    with open(args.model_file, "w") as write_file:
+        json.dump(dict_symb, write_file)
+
+
+def hack(args):
+    input_text = read_from_file(args.input_file)
+    with open(args.model_file, 'r') as read_file:
+        try:
+            model_dict = json.load(read_file)
+        except Exception:
+            sys.exit('model_file format is json')
+
+    my_text = input_text
+    input_text = parc(input_text)
+    len_key = hack_find_len_key(input_text, model_dict)
+    if len_key == 1:
+        return caesar_hack(args, my_text)
+    line_shifts = hack_find_line_shifts(args, input_text, len_key)
+
+    best_key = ''
+    min_dist = 100
+    for start_shift in range(SIZE_ALPHABET):
+        curr_key = ''
+        start_elem = chr(ord('a') + start_shift)
+        for i in range(len_key):
+            curr_key = curr_key + chr((ord(start_elem) + line_shifts[i] - ord('a')) % SIZE_ALPHABET + ord('a'))
+        args.key = curr_key
+        curr_dist = get_dist_str(model_dict, vigenere_decode(args, input_text))
+        if curr_dist < min_dist:
+            min_dist = curr_dist
+            best_key = curr_key
+    args.key = best_key
+
+    output_text = vigenere_decode(args, my_text)
+    write_to_file(args.output_file, output_text)
+
+
+def hack_find_len_key(input_text, model_dict):
```

Comment:
у меня не вышло расшифровать текст, зашифрованный твоим шифратором виженере, предварительно обученном на моём тексте, так что пожалуйста, скинь мне текст для обучения и пример, на котором ты тестировал свой код и на котором работает hack для виженере

Reply:
Здесь проблема наверно в строчке 179 (т.к. там примерные оценки)
Мой текст - первая глава из гарри поттера

Reply:
input.txt:
LONG AGO, IN A TIME FORGOTTEN, A PRETERNATURAL EVENT THREW THE SEASONS OUT OF BALANCE. IN A LAND WHERE SUMMERS CAN LAST DECADES AND WINTERS A LIFETIME, TROUBLE IS BREWING. THE COLD IS RETURNING, AND IN THE FROZEN WASTES TO THE NORTH OF WINTERFELL, SINISTER AND SUPERNATURAL FORCES ARE MASSING BEYOND THE KINGDOM’S PROTECTIVE WALL. AT THE CENTER OF THE CONFLICT LIE THE STARKS OF WINTERFELL, A FAMILY AS HARSH AND UNYIELDING AS THE LAND THEY WERE BORN TO. SWEEPING FROM A LAND OF BRUTAL COLD TO A DISTANT SUMMERTIME KINGDOM OF EPICUREAN PLENTY, HERE IS A TALE OF LORDS AND LADIES, SOLDIERS AND SORCERERS, ASSASSINS AND BASTARDS, WHO COME TOGETHER IN A TIME OF GRIM OMENS. HERE AN ENIGMATIC BAND OF WARRIORS BEAR SWORDS OF NO HUMAN METAL; A TRIBE OF FIERCE WILDLINGS CARRY MEN OFF INTO MADNESS; A CRUEL YOUNG DRAGON PRINCE BARTERS HIS SISTER TO WIN BACK HIS THRONE; AND A DETERMINED WOMAN UNDERTAKES THE MOST TREACHEROUS OF JOURNEYS. AMID PLOTS AND COUNTERPLOTS, TRAGEDY AND BETRAYAL, VICTORY AND TERROR, THE FATE OF THE STARKS, THEIR ALLIES, AND THEIR ENEMIES HANGS PERILOUSLY IN THE BALANCE, AS EACH ENDEAVORS TO WIN THAT DEADLIEST OF CONFLICTS: THE GAME OF THRONES.  MR. AND MRS. DURSLEY, OF NUMBER FOUR, PRIVET DRIVE, WERE PROUD TO SAY THAT THEY WERE PERFECTLY NORMAL, THANK YOU VERY MUCH. THEY WERE THE LAST PEOPLE YOU'D EXPECT TO BE INVOLVED IN ANYTHING STRANGE OR MYSTERIOUS, BECAUSE THEY JUST DIDN'T HOLD WITH SUCH NONSENSE.

Reply:
Текст для обучения я скинул в тг

Reply:
ок